JBang failures on Windows when symlinks are allowed in the FileSystem

There are some issues in the implementation, which does not take this mode in account. This is not a default, but it is a legitimate use-case for Windows 10.

## Runtime

```
C:\Users\Oleg\Documents\jbang>jbang gavsearch@jbangdev
[jbang] Downloading JDK 11. Be patient, this can take several minutes...
[jbang] Installing JDK 11...
[jbang] java.nio.file.FileSystemException: C:\Users\Oleg\.jbang\currentjdk: A required privilege is not held by the client.

[jbang] Creation of symbolic link failed.
[jbang] Now try creating a hard link instead of symbolic.
[jbang] Creation of hard link failed. Script must be on the same drive as $JBANG_CACHE_DIR (typically under $HOME) for hardlink creation to work. Or call the command with admin rights.
[jbang] [ERROR] Required Java version not possible to download or install. You can run with '--java 8' to force using the default installed Java.
[jbang] [ERROR] Unable to download or install JDK version 11
[jbang] Run with --verbose for more details
```

## Buildtime

```
> Task :test

TestEdit > testEditJitPackDepAndRepo(Path) FAILED
    dev.jbang.cli.ExitException at TestEdit.java:121
        Caused by: org.jboss.shrinkwrap.resolver.api.NoResolvedResultException at TestEdit.java:121
            Caused by: org.eclipse.aether.resolution.DependencyResolutionException at TestEdit.java:121
                Caused by: org.eclipse.aether.resolution.ArtifactResolutionException at TestEdit.java:121
                    Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException at TestEdit.java:121
java -classpath ^"C:\Users\Oleg\Documents\jbang\jbang\build\classes\java\test\itests\hellojar.jar^" helloworld
java -Dfoo=bar ^"-Dbar=aap^ noot^ mies^" -classpath ^"C:\Users\Oleg\AppData\Local\Temp\junit2046284907444777935\jbang\cache\jars\helloworld.java.6c29e59df0efda6a7a9783b1de6b823dc0371000e3c8d32b2b143a430c20900b.jar^" ^"-XX:ArchiveClassesAtExit=C:\Users\Oleg\AppData\Local\Temp\junit2046284907444777935\jbang\cache\jars\helloworld.java.6c29e59df0efda6a7a9783b1de6b823dc0371000e3c8d32b2b143a430c20900b.jar.jsa^" helloworld
<========-----> 64% EXECUTING [5m 2s]

TestSourcesMultipleSomeMissingFiles > testFindSourcesInMultipleFilesSymbolicLinkMissing() FAILED
    java.nio.file.FileSystemException at TestSourcesMultipleSomeMissingFiles.java:91

280 tests completed, 2 failed, 2 skipped
```

--- a/CONTRIBUTING.adoc
+++ b/CONTRIBUTING.adoc
@@ -4,6 +4,8 @@ Anyone is welcome to open issues and/or pull-requests for bugfixes, feature-requ
 
 == How to build
 
+=== Quickstart
+
 Clone the repository, cd `jbang` and run `./gradlew build`.
 There will be a warning in the output but don't despair - that is just
 Java being picky.
@@ -29,6 +31,13 @@ BUILD SUCCESSFUL in 5s
 ----
 <.> This error happens as in the junit tests we test behaviors under different environment variables and to do that we need to modify internal state in java's `System` handling of properties. Java 9+ does not like that.
 <.> Already https://github.com/stefanbirkner/system-rules/issues/64[reported] but it can't be fixed as the jvm does not allow this; but fine for `jbang` as this is just for the test runs.
+
+=== Building on Windows specifics
+
+The guidelines above should work for Windows, and this configuration is covered by the CI pipeline.
+Some tests may be skipped in your local environment if symbolic linking is not enabled for your filesystem
+OR if your test process does not have the required permissions for a directory.
+See the configuration guidelines in the link:readme.adoc#usage-on-windows[Usage on Windows] section.
 
 == Conventional Commits + squash and merge
 

--- a/readme.adoc
+++ b/readme.adoc
@@ -574,6 +574,29 @@ A side effect of running GAV as a jar, the GAV could also be a `.java` or `.jsh`
 No one would want to do that (right?) but now you know.
 ====
 
+=== Usage on Windows
+
+Some JBang commands need to create symbolic links when running on Windows.
+For example, this is required for Managing JDKs or editing the files with the `edit` command.
+To resolve the issue, on Windows 10 you will need to enable symbolic linking for your user account.
+
+* Review security considerations and best practices https://docs.microsoft.com/en-us/windows/security/threat-protection/security-policy-settings/create-symbolic-links[here].
+* Enable the `Create symbolic links` group policy setting for your user/group
+  by following the guidelines on the https://docs.microsoft.com/en-us/windows/security/threat-protection/security-policy-settings/user-rights-assignment[User Rights Assignment] page.
+
+You might also need elevated privileges to create symbolic links
+If the JBang process has no full administrative rights, the execution may result in an error like:
+
+```
+java.nio.file.FileSystemException: <CURRENT_JDK>: A required privilege is not held by the client.
+```
+
+To resolve this issue, see the options in https://answers.microsoft.com/en-us/windows/forum/windows_10-performance-winpc/how-to-fix-problem-a-required-privilege-is-not/d206b4b3-b6c2-4b0b-8629-5d4dfcb8f8d0[this answers.microsoft.com article].
+There are multiple possible resolutions,
+choose one meeting your needs.
+
+For older Windows versions, see this https://stackoverflow.com/a/24353758[StackOverflow issue].
+
 == Init templates
 
 To get started you can run `jbang init helloworld.java` and a simple java class with a static main is generated.

--- a/src/main/java/dev/jbang/dependencies/DependencyUtil.java
+++ b/src/main/java/dev/jbang/dependencies/DependencyUtil.java
@@ -198,7 +198,7 @@ public class DependencyUtil {
 				artifacts = stage.asList(MavenResolvedArtifact.class); // , RUNTIME);
 
 			} catch (RuntimeException e) {
-				throw new ExitException(1, "Could not resolve dependency", e);
+				throw new ExitException(1, "Could not resolve dependency " + it, e);
 			}
 
 			if (loggingEnabled)

--- a/src/main/java/dev/jbang/util/Util.java
+++ b/src/main/java/dev/jbang/util/Util.java
@@ -940,7 +940,12 @@ public class Util {
 		} catch (IOException e) {
 			infoMsg(e.toString());
 		}
-		infoMsg("Creation of symbolic link failed.");
+		if (Util.isWindows()) {
+			infoMsg("Creation of symbolic link failed." +
+					"For potential causes and resolutions see https://github.com/jbangdev/jbang/blob/main/readme.adoc#usage-on-windows");
+		} else {
+			infoMsg("Creation of symbolic link failed.");
+		}
 		return false;
 	}
 

--- a/src/test/java/dev/jbang/source/TestSourcesMultipleSomeMissingFiles.java
+++ b/src/test/java/dev/jbang/source/TestSourcesMultipleSomeMissingFiles.java
@@ -1,14 +1,18 @@
 package dev.jbang.source;
 
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 import java.io.IOException;
+import java.nio.file.FileSystemException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
 import java.util.List;
 import java.util.TreeSet;
 
+import org.junit.Assume;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
@@ -88,7 +92,13 @@ class TestSourcesMultipleSomeMissingFiles extends BaseTest {
 		t.toFile().delete();
 
 		target = Files.copy(target, target.getParent().resolve("C3.java"), StandardCopyOption.REPLACE_EXISTING);
-		Files.createSymbolicLink(t, target);
+		try {
+			Files.createSymbolicLink(t, target);
+		} catch (FileSystemException ex) {
+			Assume.assumeThat("Cannot check symbolic link. Permissions should be enabled first. " +
+							"See https://github.com/jbangdev/jbang/blob/main/CONTRIBUTING.adoc#building-on-windows-specifics",
+					ex.getReason(), not(containsString("A required privilege is not held by the client")));
+		}
 		Files.delete(target);
 
 		TestScript.createTmpFileWithContent(HiJBangPath.getParent(), "inner", "HelloInner.java",

--- a/src/test/java/dev/jbang/source/TestSourcesMultipleSomeMissingFiles.java
+++ b/src/test/java/dev/jbang/source/TestSourcesMultipleSomeMissingFiles.java
@@ -96,7 +96,7 @@ class TestSourcesMultipleSomeMissingFiles extends BaseTest {
 			Files.createSymbolicLink(t, target);
 		} catch (FileSystemException ex) {
 			Assume.assumeThat("Cannot check symbolic link. Permissions should be enabled first. " +
-							"See https://github.com/jbangdev/jbang/blob/main/CONTRIBUTING.adoc#building-on-windows-specifics",
+					"See https://github.com/jbangdev/jbang/blob/main/CONTRIBUTING.adoc#building-on-windows-specifics",
 					ex.getReason(), not(containsString("A required privilege is not held by the client")));
 		}
 		Files.delete(target);

